qgis: Updates for proj@6

citibeth opened the pull request:
* Added dependency libtasn1, previously missed.  (tasn1 is a low-level GNU library)

* Updates to support proj@6:
  * Updated to grass@7.8.1: for +grass7 because it support proj@6
  * Updated dependency to proj@6, depending on QGIS version (QGIS docs are unclear about which QGIS version is the first to support proj@6)

* Removed dependency on HDF5.  It's not needed because netcdf-c already depends on HDF5.  Its presence causes concretization errors.

Commit:
qgis

adamjstewart commented on var/spack/repos/builtin/packages/qgis/package.py:
```diff
@@ -97,10 +98,11 @@ class Qgis(CMakePackage):
     depends_on('postgresql@8:', when='+postgresql')  # for PostGIS support
     depends_on('gsl', when='+georeferencer')  # for georeferencer
-    depends_on('grass@7.0.0', type=('build', 'link', 'run'), when='+grass7')  # for georeferencer
+    # grass@7.8.1 is the first version that supports proj@6
```
But this package still supports PROJ 4, right? In that case, 7.8.1 isn't required.

citibeth replied:
No it does not support `proj@4`.  I was not able to get `qgis` to concretize and build with `proj@4`.  That why I went to `proj@6` and updated a zillion packages to try to get them working.

Sinan81 replied:
I remember building it with `proj@4` up to QGIS@3.8.1 inclusive. ~~LTR version, @3.4 though must have dropped the support at some point.~~

Sinan81 replied:
Even in the latest version, it seems QGIS supports proj@4 unless one wants to build with grass@7.

citibeth replied:
> Even in the latest version, it seems QGIS supports proj@4 unless one wants to build with grass@7.

Thanks for that info.  Does it work with any versions of GRASS before 7?  I suppose GRASS is optional, but I believed I needed it...

I think the PROJ.4/6 thing is going to be like Python 2/3: It's giving us a lot of pain right now, but in the next 1-2 years, things will have pretty much converted over and we won't be worrying about it anymore.

adamjstewart commented on var/spack/repos/builtin/packages/qgis/package.py:
```diff
@@ -97,10 +98,11 @@ class Qgis(CMakePackage):
 
     # the below dependencies are shown in cmake config
-    depends_on('hdf5')
+#    depends_on('hdf5')
```
If this package directly links to hdf5, I would still keep this dependency. Explicit is better than implicit. In the past, we've considered removing transitive dependencies from the compiler wrappers, but they broke a lot of packages, so we undid that change.

citibeth replied:
It breaks the concretizer.  `Spack spec qgis` results in a spectacular error.

adamjstewart replied:
Ugh, can't wait for that new concretizer...

citibeth replied:
Until then, do you think we could do what's needed to ensure that things concretize?  Having specs that simply don't concretize out of the box is bad PR, IMHO.

adamjstewart replied:
Yeah, I agree, this is fine.

adamjstewart commented on var/spack/repos/builtin/packages/qgis/package.py:
```diff
@@ -111,6 +113,7 @@ class Qgis(CMakePackage):
 
     # Take care of conflicts using depends_on
     depends_on('proj@5:', when='@3.8.2:')
+    depends_on('proj@6.3.1:', when='@3.12.1:')    # Not sure of the when= version.
```
@Sinan81 might know where to find the exact version requirements.

Sinan81 replied:
For dependencies I check out either of the following documents which are retrieved from branch `final-3_12_1` for example:

- https://github.com/qgis/QGIS/blob/final-3_12_1/INSTALL
- http://htmlpreview.github.io/?https://raw.github.com/qgis/QGIS/final-3_12_1/doc/INSTALL.html

every release has a corresponding `final-...` tag:
```
$ git clone https://github.com/qgis/QGIS.git
$ cd QGIS
$ git tag
...
final-3_12_2
final-3_2_0
final-3_2_1
final-3_2_2
final-3_2_3
final-3_4_0
final-3_4_1
final-3_4_10
final-3_4_11
final-3_4_12
final-3_4_13
final-3_4_14
final-3_4_15
final-3_4_2
final-3_4_3
final-3_4_4
final-3_4_5
final-3_4_6
final-3_4_7
final-3_4_8
final-3_4_9
final-3_6_0
final-3_6_1
final-3_6_2
final-3_6_3
final-3_8_0
final-3_8_1
final-3_8_2
final-3_8_3
...
```

Sinan81 commented:
FYI, this is the command I use to give spack enough hints that It will move on to installation right away and won't complain about dependency resolution.
```
spack install qgis ^gdal@2.4.2 ^ncurses+termlib ^mesa~llvm ^proj@5.2.0 ^libgeotiff@1.4.3 ^netcdf-c+mpi ^py-pyproj@1.9.6 ^python@3.7.4
```
A few of the hints above serve to constraine proj version to 5. netcdf+mpi takes care of hdf5 concretization error. mesa~llvm is possibly optional (but necessary if gcc version is old)

Sinan81 commented:
It seems `libspatialite` is the limiting factor in terms of building QGIS with proj@6: latest release only supports proj@:5. The current `@5.x.x-beta` release seem to support proj@6. In the following ticket it is mentioned that proj@6 is fully supported as of a commit about a year ago: https://www.gaia-gis.it/fossil/libspatialite/info/6eabdf0e6d3520a7

citibeth commented:
> FYI, this is the command I use to give spack enough hints that It will move on to installation right away and won't complain about dependency resolution.
> 
> ```
> spack install qgis ^gdal@2.4.2 ^ncurses+termlib ^mesa~llvm ^proj@5.2.0 ^libgeotiff@1.4.3 ^netcdf-c+mpi ^py-pyproj@1.9.6 ^python@3.7.4
> ```
> 
> A few of the hints above serve to constraine proj version to 5. netcdf+mpi takes care of hdf5 concretization error. mesa~llvm is possibly optional (but necessary if gcc version is old)

This is nice to know, but it's not in any place that will be useful to someone trying to build QGIS in the future.  Where can we put this information?  Also... for me, it's something I would put in a `spack.yaml` file.

citibeth commented:
> It seems `libspatialite` is the limiting factor in terms of building QGIS with proj@6: latest release only supports proj@:5. The current `@5.x.x-beta` release seem to support proj@6. In the following ticket it is mentioned that proj@6 is fully supported as of a commit about a year ago: https://www.gaia-gis.it/fossil/libspatialite/info/6eabdf0e6d3520a7

That's right, see here: https://github.com/spack/spack/pull/16173/files

citibeth committed:
Update package.py

QGIS 3.12.1 can use PROJ >= 4.9.3.  Therefore both version restrictions on PROJ were incorrect.
https://github.com/qgis/QGIS/blob/final-3_12_1/INSTALL

citibeth committed:
Update package.py

Add explanation to (hopefully temporary) removal of hdf5 dependency.

citibeth committed:
Remove overly restrictive GRASS version number.

citibeth commented:
OK.... I haven't been able to compile QGIS (with PROJ.6 or any other) on macOS.  But the changes in this PR seem to be an improvement over what was there before, in terms of more correclty matching published requirements.

citibeth committed:
flake8

Sinan81 commented:
Looks good to me. I wonder if @adamjstewart tried compiling it on MacOS?

adamjstewart commented:
Nope, haven't needed QGIS in a while

adamjstewart commented:
Closing and reopening to reboot the tests...
